### PR TITLE
Validators refactoring

### LINKED ISSUE #351: Allow customization of validator / Apply defaults in request body
### Description

I need Connexion to use default value if a property is missing in request body. Although it is done with URL query, in body 'default' is just for documentation purposes (that is nature of JSON Schema). In order to enforce defaults (as described here https://python-jsonschema.readthedocs.io/en/latest/faq/#why-doesn-t-my-schema-that-has-a-default-property-actually-set-the-default-on-my-instance) I need to overwrite validator used in `connexion/decorators/validation.py`. It would be even better if Connexion offered defaults enforcement on its own.

### Expected bahaviour

- Access to body validation class in similar way as we have it with Resolver.

OR

- A flag to set 'defaults enforcement'. When 'defaults' is defined for a non-required property and the property is not passed, default value should be used.

### Actual behaviour

We don't have tools for that. When non-required property is not passed, it is not included in data provided for controlled function, even if there are defaults.

### Example

```yaml
...
    post:
      operationId: "controllers.post_foo"
      parameters:
      - in: "body"
        name: "foo"
        required: true
        schema:
          $ref: "#/definitions/Foo"
      responses:
        200:
          description: "OK"
definitions:
  Foo:
    type: "object"
    properties:
      name:
        type: "string"
        default: "BOB"
```

If `-d '{}'` was sent, controller should receive `{'name': 'BOB'}`.

### Another example

It should be enough to provide defaults in bottom-most layer.

```yaml
definitions:
  Foo:
    type: "object"
    default: {}
    properties:
      name:
        type: "object"
        properties:
          first:
            type: "string"
            default: "BOB"
          last:
            type: "string"
            default: "DOE"
```

If `-d '{}'` was sent, controller should receive `{'name': {'first': 'BOB', 'last': 'DOE'}}`.

### PR DESCRIPTION
**Pull request by phoracek**
Note that this must be merged only after #281.

Changes proposed in this pull request:

 - Drop unused attribute from ResponseBodyValidator
 - Use Draft4Validator instance directly instead of validation() shorthand
 - Allow user to override default Validator used to validate request and response body, thanks to this, it is easier to customize existing Validators

### CODE REVIEW

**Comment by phoracek** [comment]
Sorry, I totally missed https://github.com/zalando/connexion/pull/281.

I think we can use first two patches from this pull request and maybe also slightly changed examples and cookbook.

**Comment by rafaelcaricio** [comment]
@phoracek Would you be willing to write a new PR for documenting the #281 and maybe changing something you think would fit some additional idea from your proposal? Sorry, I didn't had time to reply to you on #351 before you started coding your PR.

**Comment by phoracek** [comment]
@rafaelcaricio I pushed promised changes, but bear in mind, that it needs to be merged after your PR. What should I do now? Rebase this PR on your, or wait for yours to be merged or rebased on master?

**Comment by phoracek** [comment]
I rebased original custom validators PR #281  on top of master and set my on top of that. Hope it helps.

**Comment by rafaelcaricio** [comment]
@phoracek Thank you very much for the work in this PR. It looks good to me.

We just have to isort the imports, applying this patch:

```patch
--- /home/travis/build/zalando/connexion/examples/enforcedefaults/enforcedefaults.py:before	2016-12-03 17:44:15.006521
+++ /home/travis/build/zalando/connexion/examples/enforcedefaults/enforcedefaults.py:after	2016-12-03 17:47:19.644362
@@ -1,9 +1,9 @@
 #!/usr/bin/env python3
 
-from connexion.decorators.validation import RequestBodyValidator
 import connexion
 import jsonschema
 import six
+from connexion.decorators.validation import RequestBodyValidator
```

**Comment by phoracek** [comment]
Thanks Rafael :) I resolved the isort problem. Example works, so I hope it won't break anything.

**Comment by rafaelcaricio** [comment]
👍

File: examples/enforcedefaults/README.rst [hunk 1]
```diff
@@ -0,0 +1,18 @@
...
+Validator code is based on example from `python-jsonschema docs`_.
+
+Running:
+
+.. code-block:: bash
+
+    $ ./enforcedefaults.py
+
+Now open your browser and go to http://localhost:8080/v1/ui/ to see the Swagger
+UI. If you send a ``POST`` request with empty body ``{}``, you should recieve
```
Comment (rafaelcaricio) [inline examples/enforcedefaults/README.rst]: typo: s/recieve/receive/

File: docs/request.rst [hunk 2]
```diff
@@ -148,3 +148,22 @@ Currently, header parameters are not passed to the handler functions as paramete
 
     def index():
         page_number = connexion.request.headers['Page-Number']
+
+
+Custom Validators
+-----------------
+
+By default, body and parameters content is validated againts OpenAPI scheme
```
Comment (rafaelcaricio) [inline docs/request.rst]: s/is/are/
s/againts/against/
s/scheme/schema/

File: docs/request.rst [hunk 3]
```diff
@@ -148,3 +148,22 @@ Currently, header parameters are not passed to the handler functions as paramete
...
+via ``connexion.decorators.validation.RequestBodyValidator``
+or ``connexion.decorators.validation.ParameterValidator``, if you want to
+change validation, you can override defaults with:
```
Comment (rafaelcaricio) [inline docs/request.rst]: "change **the** validation, you can override **the** defaults with"

File: examples/enforcedefaults/README.rst (hunk 1 above)
Comment (phoracek) [inline examples/enforcedefaults/README.rst]: Fixed, always miss this one.

File: docs/response.rst [hunk 4]
```diff
@@ -66,6 +66,22 @@ do so by opting in when adding the API:
 This will validate all the responses using `jsonschema` and is specially useful
 during development.
 
+
+Custom Validator
+-----------------
+
+By default, response body content is validated againts OpenAPI scheme
```
Comment (rafaelcaricio) [inline docs/response.rst]: s/is/are/
s/againts/against/
s/scheme/schema/

File: docs/response.rst [hunk 5]
```diff
@@ -66,6 +66,22 @@ do so by opting in when adding the API:
...
+via ``connexion.decorators.response.ResponseValidator``, if you want to change
+validation, you can override default class with:
```
Comment (rafaelcaricio) [inline docs/response.rst]: "..change **the** validation, you can override **the** default class with..."

**Comment by rafaelcaricio** [comment]
👍

**Pull request merged**